dali/pipeline/operators/arithmetic/expression_tree.h

```diff
@@ -0,0 +1,240 @@
+
+namespace dali {
+
+/**
+ * @brief Describe a tile of data to be processed in expression evaluation.
+ */
+struct TileDesc {
+  int sample_idx;       // id of sample inside within the batch
+  int extent_idx;       // the index of tile within this sample_idx
+  int64_t extent_size;  // actually covered extent in this tile, the last can be smaller
+  int64_t tile_size;    // the size of regular tile
+};
+
+inline std::ostream &operator<<(std::ostream &os, const TileDesc &v) {
+  os << "{" << v.sample_idx << ", " << v.extent_idx << ", " << v.extent_size << ", " << v.tile_size
+     << "}";
+  return os;
+}
+
+struct TileRange {
+  int begin;
+  int end;
+};
+
+inline std::ostream &operator<<(std::ostream &os, const TileRange &v) {
+  os << "{" << v.begin << ", " << v.end << "}";
+  return os;
+}
+
+class ExprNode;
+
+struct ExpressionImplContext {
+  const ExprNode *node;
+};
+
+/**
+ * @brief Part responsible for loop-execution over tile, the implementation knows the types
+ * and can access the inputs internally
+ */
+class ExpressionImplBase {
+ public:
+  virtual void Execute(ArgumentWorkspace &workspace, const OpSpec &spec, ExpressionImplContext &ctx,
+                       const std::vector<TileDesc> &tiles, TileRange range) = 0;
+  virtual ~ExpressionImplBase() = default;
+};
+
+enum class NodeType { Function, Constant, Tensor };
+
+inline std::string GetAbbreviation(NodeType t) {
+  switch (t) {
+    case NodeType::Function:
+      return "F";
+    case NodeType::Constant:
+      return "C";
+    case NodeType::Tensor:
+      return "T";
+    default:
+      DALI_FAIL("Unrecognized NodeType.");
+  }
+}
+
+/**
+ * @brief base class for nodes of ExpressionTree
+ *
+ * TODO(klecki): Using unique_ptrs for ownership, we may generalize it to DAG, that would require
+ * some changes
+ */
+class ExprNode {
+ public:
+  virtual const std::string &GetOp() const {
+    DALI_FAIL("No op in this expression node.");
+  }
+
+  virtual std::string GetNodeDesc() const {
+    return GetOutputDesc();
+  }
+
+  virtual std::string GetOutputDesc() const {
+    auto op_type = TypeTable::GetTypeInfo(GetTypeId()).name();
+    std::string result = GetAbbreviation(GetNodeType());
+    result += IsScalarLike(GetShape()) ? "C:" : "T:";
+    return result + op_type;
+  }
+
+  virtual NodeType GetNodeType() const = 0;
+
+  std::unique_ptr<ExpressionImplBase> impl;
+
+  void SetTypeId(DALIDataType type_id) {
+    type_id_ = type_id;
+  }
+
+  DALIDataType GetTypeId() const {
+    return type_id_;
+  }
+
+  void SetShape(const kernels::TensorListShape<> &shape) {
+    shape_ = shape;
+  }
+
+  const kernels::TensorListShape<> &GetShape() const {
+    return shape_;
+  }
+
+  virtual int GetSubexpressionCount() const {
+    return 0;
+  }
+
+  virtual ExprNode &operator[](int i) {
+    DALI_FAIL("No subexpressions for this node.");
+  }
+
+  virtual const ExprNode &operator[](int i) const {
+    DALI_FAIL("No subexpressions for this node.");
+  }
```

Comment:
Unless you make it pure virtual (which is poor idea I guess), you should provide a clear data structure to contain subexpression in this class. It's because different subclasses could implement it in different ways, and it might lead to crazy surprises, e.g. in complexity. Take a look at this code:
```
ExprNode* a = new LinAccessExprNode();
ExprNode* b = new LognAccessExprNode();

*a[0];  // linear time
*b[0];  // logn time
```

This being said, I'd also think about, why should this be virtual. Is it necessary here?